### PR TITLE
test: migrate patch, pkgid, proc_macro and progress to snapbox

### PR DESCRIPTION
**Pull request by eth3lbert**
### What does this PR try to resolve?

Part of https://github.com/rust-lang/cargo/issues/14039.

Migrate following to snapbox:

- `tests/testsuite/patch.rs`
- `tests/testsuite/pkgid.rs`
- `tests/testsuite/proc_macro.rs`
- `tests/testsuite/progress.rs`

### CODE REVIEW

File: tests/testsuite/pkgid.rs [hunk 1]
```diff
@@ -311,75 +307,27 @@ fn pkgid_json_message_metadata_consistency() {
...
-        )
+            .with_stdout_data(str![[r#"
+{"executable":null,"features":[],"filenames":"{...}","fresh":false,"manifest_path":"[ROOT]/foo/Cargo.toml","package_id":"path+[ROOTURL]/foo#0.5.0","profile":"{...}","reason":"compiler-artifact","target":"{...}"}
+{"cfgs":[],"env":[],"linked_libs":[],"linked_paths":[],"out_dir":"[ROOT]/foo/target/debug/build/foo-[HASH]/out","package_id":"path+[ROOTURL]/foo#0.5.0","reason":"build-script-executed"}
+{"manifest_path":"[ROOT]/foo/Cargo.toml","message":"{...}","package_id":"path+[ROOTURL]/foo#0.5.0","reason":"compiler-message","target":"{...}"}
+{"manifest_path":"[ROOT]/foo/Cargo.toml","message":"{...}","package_id":"path+[ROOTURL]/foo#0.5.0","reason":"compiler-message","target":"{...}"}
+{"executable":null,"features":[],"filenames":"{...}","fresh":false,"manifest_path":"[ROOT]/foo/Cargo.toml","package_id":"path+[ROOTURL]/foo#0.5.0","profile":"{...}","reason":"compiler-artifact","target":"{...}"}
+{"reason":"build-finished","success":true}
+
+"#]].json_lines())
```
Comment (eth3lbert) [inline tests/testsuite/pkgid.rs:318]: Please let me know if there's a better approach to working with prettified json.

Comment (epage) [reply tests/testsuite/pkgid.rs:318]: Probably better to not do `json_lines` at this time.  We want to find a better solution for pretty printing it.

File: tests/testsuite/progress.rs [hunk 2]
```diff
@@ -156,9 +162,28 @@ fn never_progress() {
...
 
+    // should not contain
+    // [DOWNLOADING] [..] crates [..]
+    // [..][DOWNLOADED] 3 crates ([..]) in [..]
+    // [BUILDING] [..] [..]/4: [..]
     p.cargo("check")
-        .with_stderr_does_not_contain("[DOWNLOADING] [..] crates [..]")
-        .with_stderr_does_not_contain("[..][DOWNLOADED] 3 crates ([..]) in [..]")
-        .with_stderr_does_not_contain("[BUILDING] [..] [..]/4: [..]")
+        .with_stderr_data(
```
Comment (epage) [inline tests/testsuite/progress.rs:170]: Let's revert this for now.  People are unlikely to see these comments when updating snapshots

Comment (eth3lbert) [reply tests/testsuite/progress.rs]: Revered!

File: tests/testsuite/pkgid.rs (hunk 1 above)
Comment (eth3lbert) [reply tests/testsuite/pkgid.rs]: I've reverted this change and allowed the deprecated `with_json` for now.

**Pull request merged**